pkg/router/metrics/haproxy/haproxy.go

```diff
@@ -0,0 +1,653 @@
+			8:  newServerMetric("bytes_in_total", "Current total of incoming bytes.", nil),
+			9:  newServerMetric("bytes_out_total", "Current total of outgoing bytes.", nil),
+			13: newServerMetric("connection_errors_total", "Total of connection errors.", nil),
+			14: newServerMetric("response_errors_total", "Total of response errors.", nil),
+			15: newServerMetric("retry_warnings_total", "Total of retry warnings.", nil),
+			16: newServerMetric("redispatch_warnings_total", "Total of redispatch warnings.", nil),
+			17: newServerMetric("up", "Current health status of the server (1 = UP, 0 = DOWN).", nil),
+			18: newServerMetric("weight", "Current weight of the server.", nil),
+			21: newServerMetric("check_failures_total", "Total number of failed health checks.", nil),
+			24: newServerMetric("downtime_seconds_total", "Total downtime in seconds.", nil),
+			33: newServerMetric("current_session_rate", "Current number of sessions per second over last elapsed second.", nil),
+			35: newServerMetric("max_session_rate", "Maximum observed number of sessions per second.", nil),
+			38: newServerMetric("check_duration_milliseconds", "Previously run health check duration, in milliseconds", nil),
+			39: newServerMetric("http_responses_total", "Total of HTTP responses.", prometheus.Labels{"code": "1xx"}),
+			40: newServerMetric("http_responses_total", "Total of HTTP responses.", prometheus.Labels{"code": "2xx"}),
+			41: newServerMetric("http_responses_total", "Total of HTTP responses.", prometheus.Labels{"code": "3xx"}),
+			42: newServerMetric("http_responses_total", "Total of HTTP responses.", prometheus.Labels{"code": "4xx"}),
+			43: newServerMetric("http_responses_total", "Total of HTTP responses.", prometheus.Labels{"code": "5xx"}),
+			44: newServerMetric("http_responses_total", "Total of HTTP responses.", prometheus.Labels{"code": "other"}),
+			60: newServerMetric("http_average_response_latency_milliseconds", "Average response latency of the last 1024 requests in milliseconds.", nil),
+		}),
+	}, nil
+}
+
+// Describe describes all the metrics ever exported by the HAProxy exporter. It
+// implements prometheus.Collector.
+func (e *Exporter) Describe(ch chan<- *prometheus.Desc) {
+	for _, m := range e.frontendMetrics {
+		m.Describe(ch)
+	}
+	for _, m := range e.backendMetrics {
+		m.Describe(ch)
+	}
+	for _, m := range e.serverMetrics {
+		m.Describe(ch)
+	}
+	ch <- e.up.Desc()
+	ch <- e.totalScrapes.Desc()
+	ch <- e.nextScrapeInterval.Desc()
+	ch <- e.serverThresholdCurrent.Desc()
+	ch <- e.serverThresholdLimit.Desc()
+	ch <- e.csvParseFailures.Desc()
+}
+
+// Collect fetches the stats from configured HAProxy location and delivers them
+// as Prometheus metrics. It implements prometheus.Collector. It will not collect
+// metrics more often than interval.
+func (e *Exporter) Collect(ch chan<- prometheus.Metric) {
+	e.mutex.Lock()
+	defer e.mutex.Unlock()
+
+	now := time.Now()
+	switch {
+	case e.pendingScrape:
+		// CollectNow was called before
```

Comment:
Example would be total bytes from a server (what is captured).  A sampler grabs those and calculates a derivative over them (we'd probably have to handle wraparound, but there are tools for that).  The derivative of bytes/sec is useful no matter what.